src/librustc/hir/def_id.rs

```diff
@@ -84,8 +85,16 @@ impl fmt::Display for CrateNum {
     }
 }
 
-impl rustc_serialize::UseSpecializedEncodable for CrateNum {}
-impl rustc_serialize::UseSpecializedDecodable for CrateNum {}
+impl rustc_serialize::UseSpecializedEncodable for CrateNum {
+    fn default_encode<E: Encoder>(&self, e: &mut E) -> Result<(), E::Error> {
+        e.emit_u32(self.as_u32())
```

Comment:
Thank you for bringing this up. I was actually struggled a lot trying to emit the correct data. I believe we can safely simply emit the crate number here.
As you pointed out, the crate numbers is only meaningful within the same compilation session; therefore, one must have the session info when using crate numbers. For example, one might need include a crate number -> `Svh` mapping table in the serialized data, or some other ways to correctly map crate numbers to the actual crates.

In this PR, crate numbers are used within one `CodegenResults`, and `CodegenResults` only belongs to one specific session; therefore, crate numbers are unique. Then we can use `CodegenResults.crate_info.used_crate_source` to find the crate from a crate number.

Reply:
You also replaced the `Encodable` impl of `CrateNum` for other usages. Maybe make a new `LinkingCrateNum` newtype for `u32` and store that in `crate_info` instead?

Reply:
I don't think this `Encodable` impl is currently used since the [default impl](https://github.com/rust-lang/rust/blob/9ff30a7810c586819a78188c173a7b74adbb9730/src/libserialize/serialize.rs#L889-L894) just returns an error.
However, I like your idea of making a new `LinkingCrateNum` as it conceptually decouples `CrateInfo` from a compilation session. I'll update accordingly.

Reply:
I created #67516 for decoupling `CrateNum` from the compilation session. @bjorn3 could you take a look?